ql/lib/codeql/kaleidoscope/Cfg.qll

```diff
@@ -0,0 +1,254 @@
+private import codeql.kaleidoscope.Ast
+private import codeql.controlflow.Cfg as CfgShared
+private import codeql.Locations
+private import codeql.kaleidoscope.controlflow.BasicBlocks
+
+private module Completion {
+  private newtype TCompletion =
+    TSimpleCompletion() or
+    TBooleanCompletion(boolean b) { b in [false, true] } or
+    TReturnCompletion()
+
+  abstract class Completion extends TCompletion {
+    abstract string toString();
+
+    predicate isValidForSpecific(AstNode e) { none() }
+
+    predicate isValidFor(AstNode e) { this.isValidForSpecific(e) }
+
+    abstract SuccessorType getAMatchingSuccessorType();
+  }
+
+  class SimpleCompletion extends Completion, TSimpleCompletion {
```

Comment:
Normally it is convenient to have a `NormalCompletion` base class that both `SimpleCompletion` and `BooleanCompletion` extend.